### ISSUE TITLE
ImportRunBasedPurgeStrategy performance issue

### ISSUE THREAD

**Issue by lazyfrosch** [Icinga/icingaweb2-module-director#1422]
*Fair warning:* In the depths of director.

While working on bigger sync jobs I noticed that Sync runs can take a long of time without making sense at all.

I tracked this down to `ImportRunBasedPurgeStrategy` so far, culprit seems to be this query:

https://github.com/Icinga/icingaweb2-module-director/blob/e9a570e96d77f7f083ba5f366b94ceb09582975f/library/Director/Import/PurgeStrategy/ImportRunBasedPurgeStrategy.php#L55-L67

With my test data I want to import around 6k hosts, with only some hosts present currently.

The execution of this query takes around 45-55 seconds!

With some benchmarking, these are the raw numbers for 2 single, and the combined query.

```
Purge: Begin finding keys	824.336	2.211	22.60 KiB	13.28 MiB
Query A: 7032	835.341	11.005	3.32 MiB	16.60 MiB
Query B: 7032	847.241	11.900	3.32 MiB	19.92 MiB
Query Sub-Select: 0	51325.518	50478.277	496.00 B	19.92 MiB
```

I have no idea what is going on with the database server: `Ver 15.1 Distrib 10.2.11-MariaDB`

**Comment by lazyfrosch** [comment]
Crude debug snippet before the `fetchCol()` call:

```php
        echo $selectA;
        $a = $db->fetchAll($selectA);
        Benchmark::measure('Query A: ' . count($a));

        echo $selectB;
        $b = $db->fetchAll($selectB);
        Benchmark::measure('Query B: '. count($b));

        echo $query;
        $all = $db->fetchAll($query);
        Benchmark::measure('Query Sub-Select: ' . count($all));

        echo Benchmark::dump();

        echo Util::wantHtml($db->fetchAll('EXPLAIN ' . $query))->render();

        exit;
```

**Comment by Thomas-Gelf** [comment]
Can you show me the `EXPLAIN` output? 6000 objects shoultn'd hurt at all. And please also:

```sql
SELECT COUNT(*) FROM imported_property;
SELECT COUNT(*) FROM imported_row_property;
SELECT COUNT(*) FROM imported_rowset_row;
SELECT COUNT(*) FROM imported_property;
```

**Comment by Thomas-Gelf** [comment]
Oh, and your `innodb_buffer_pool_size` would be interesting.

**Comment by lazyfrosch** [comment]
Okay let's come back to this, output of EXPLAIN looks like (had to run it through PHP, bc of binary keys).

The dataset is not that small, but somehow MariaDB is doing weird stuff with the query.

```json
{
    "id": "1",
    "select_type": "SIMPLE",
    "table": "rsr",
    "type": "ref",
    "possible_keys": "PRIMARY,imported_rowset_row_row",
    "key": "PRIMARY",
    "key_len": "22",
    "ref": "const",
    "rows": "11421",
    "extra": "Using where; Using index"
},
{
    "id": "1",
    "select_type": "SIMPLE",
    "table": "r",
    "type": "eq_ref",
    "possible_keys": "PRIMARY",
    "key": "PRIMARY",
    "key_len": "22",
    "ref": "director.rsr.row_checksum",
    "rows": "1",
    "extra": ""
},
{
    "id": "1",
    "select_type": "SIMPLE",
    "table": "rsr",
    "type": "ref",
    "possible_keys": "PRIMARY,imported_rowset_row_row",
    "key": "PRIMARY",
    "key_len": "22",
    "ref": "const",
    "rows": "1427",
    "extra": "Using where; Using index"
},
{
    "id": "1",
    "select_type": "SIMPLE",
    "table": "r",
    "type": "eq_ref",
    "possible_keys": "PRIMARY",
    "key": "PRIMARY",
    "key_len": "22",
    "ref": "director.rsr.row_checksum",
    "rows": "1",
    "extra": "Using where; Not exists"
}
```

Database counters:

```
MariaDB [director]> SELECT COUNT(*) FROM imported_property;
+----------+
| COUNT(*) |
+----------+
|    59260 |
+----------+
1 row in set (0.06 sec)

MariaDB [director]> SELECT COUNT(*) FROM imported_row_property;
+----------+
| COUNT(*) |
+----------+
|   224325 |
+----------+
1 row in set (0.07 sec)

MariaDB [director]> SELECT COUNT(*) FROM imported_rowset_row;
+----------+
| COUNT(*) |
+----------+
|    22978 |
+----------+
1 row in set (0.03 sec)

MariaDB [director]> SELECT COUNT(*) FROM imported_property;
+----------+
| COUNT(*) |
+----------+
|    59260 |
+----------+
1 row in set (0.00 sec)

MariaDB [director]> show variables like 'innodb_buffer_pool_size';
+-------------------------+-----------+
| Variable_name           | Value     |
+-------------------------+-----------+
| innodb_buffer_pool_size | 268435456 |
+-------------------------+-----------+
1 row in set (0.01 sec)
```

**Comment by lazyfrosch** [comment]
Maybe we should just do 2 selects and sort out in PHP?

**Comment by Thomas-Gelf** [comment]
InnoDB buffer pool isn't that large, but the shown data is also relatively small. Tasks shown by explain should be cheap. I never experienced this, even when dealing with much larger data sets. Very strange.

**Comment by lazyfrosch** [comment]
MariaDB is basically running on defaults in a container. This is so weird, because of the very simple query.

Maybe a problem with binary keys of some sort?

**Comment by Thomas-Gelf** [comment]
Not really. Seems that the problem is that your MariaDB (which version are you running?) is going for the wrong strategy, the LEFT join might be the culprit, but we cannot change this. Please try to just fire the two distinct queries with `fetchCol` and then `array_diff` the result. If your shown numbers are correct, this should be much faster.

**Comment by lazyfrosch** [comment]
See end of top comment, this also fire both queries for testing, same contents, 0 diff. Also version is there :wink:

**Comment by Thomas-Gelf** [comment]
Oh, I see. 10.2 should be pretty good.

**Comment by Thomas-Gelf** [comment]
I have been able to reproduce this on MariaDB 10.0 - will be fixed with #1475, thanks a lot for debugging this!
